runtime/runtime-params-estimator/src/trie.rs

```diff
@@ -0,0 +1,417 @@
+    let (cost_short_key, ext_cost_short_key) = aggregate_per_block_measurements(
+        testbed.config,
+        1,
+        short_key_results[warmup_iters..].to_vec(),
+    );
+    let (cost_long_key, ext_cost_long_key) = aggregate_per_block_measurements(
+        testbed.config,
+        1,
+        long_key_results[warmup_iters..].to_vec(),
+    );
+    let nodes_touched_delta = ext_cost_long_key[&ExtCosts::touching_trie_node]
+        - ext_cost_short_key[&ExtCosts::touching_trie_node];
+    // The exact number of touched nodes is a implementation that we don't want
+    // to test here but it should be close to 2*final_key_len
+    assert!(nodes_touched_delta as usize <= 2 * final_key_len + 10);
+    assert!(nodes_touched_delta as usize >= 2 * final_key_len - 10);
+    let cost_delta =
+        cost_long_key.saturating_sub(&cost_short_key, &NonNegativeTolerance::PER_MILLE);
+    let cost = cost_delta / nodes_touched_delta;
+    cost
+}
+
+pub(crate) fn read_node_from_chunk_cache(testbed: &mut Testbed) -> GasCost {
+    let debug = testbed.config.debug;
+    let iters = 200;
+    let percentiles_of_interest = &[0.5, 0.9, 0.99, 0.999];
+
+    // Worst-case
+    // - L3 CPU cache is filled with dummy data before measuring
+    let spoil_l3 = true;
+    // - Completely cold cache
+    let num_warmup_values = 0;
+    // - Single node read, no amortization possible
+    let num_values = 1;
+    // - Data is spread in main memory
+    let data_spread_factor = 7;
+
+    // For the base case, worst-case assumption is slightly relaxed. The base at
+    // the 90th percentile case is used as final estimation.
+    let base_case = {
+        // Reading a different node before the measurement loads data structure
+        // into cache. It would be difficult for an attacker to avoid this
+        // consistently, so the base case assumes this is in cache.
+        let num_warmup_values = 1;
+        // Some amortization should also be allowed, or how would an attacker
+        // actually abuse undercharged costs?
+        let num_values = 16;
+
+        let results = read_node_from_chunk_cache_ext(
+            testbed,
+            iters,
+            num_values,
+            num_warmup_values,
+            data_spread_factor,
+            spoil_l3,
+        );
+        let mut p_results = percentiles(results, percentiles_of_interest);
+        if debug {
+            eprintln!(
+                "{:<32}{:>8.3} {:>8.3} {:>8.3} {:>8.3}",
+                "",
+                percentiles_of_interest[0],
+                percentiles_of_interest[1],
+                percentiles_of_interest[2],
+                percentiles_of_interest[3]
+            );
+            eprint!("{:<32}", "Base Case");
+            for cost in p_results.iter() {
+                eprint!("{:>8} ", cost.to_gas() / 1_000_000);
+            }
+            eprintln!();
+        }
+        // Take the 90th percentile measured.
+        p_results.swap_remove(1)
```

Comment:
It looks to me like this should be put in its own function.  We’re doing the same thing over and over only with the parameters changing.  Could be local function or even local closure so you wouldn’t need to pass percentiles_of_interest and testbed as arguments.

Reply:
Yeah I've put it in a closure, looks much better now IMO